liquidity/threshold_rule.go

```diff
@@ -59,3 +62,115 @@ func (r *ThresholdRule) validate() error {
+	// that we can loop in to improve our outbound liquidity.
+	case inbound.hasSurplus() && outbound.hasDeficit():
+		amt := calculateSwapAmount(balances.capacity, outbound, inbound)
+		in := swap.TypeIn
+		return amt, &in
+
+	// If we have enough inbound and enough outbound, we do not need to
+	// take any actions at present.
+	default:
+		return 0, nil
+	}
+}
+
+// balanceRequirement describes liquidity in one direction.
+type balanceRequirement struct {
+	// currentAmount is the amount of liquidity have have in the direction.
+	currentAmount btcutil.Amount
+
+	// minimumAmount is the minimum amount of liquidity that we require in
+	// this direction.
+	minimumAmount btcutil.Amount
+}
+
+func (i balanceRequirement) hasSurplus() bool {
+	return i.currentAmount > i.minimumAmount
+}
+
+func (i balanceRequirement) hasDeficit() bool {
+	return i.currentAmount < i.minimumAmount
+}
+
+// calculateSwapAmount calculates the amount of capacity that we need to shift
+// to improve the balance of our channels, and returns a reason which provides
+// additional context. This function is used in the case where we have surplus
+// liquidity in one direction, and deficit in another. This is the case in which
+// we recommend swaps, if we have deficits on both sides, we cannot swap without
+// further unbalancing, and if we have surplus on both sides, we do not need to.
+//
+// This function calculates the portion of our total capacity we should shift
+// from the surplus side to deficit side without unbalancing the surplus side.
+// This is important, because we do not want to recommend a swap in one
+// direction that just results in our needing to produce a swap in the other
+// direction.
+func calculateSwapAmount(capacity btcutil.Amount, deficitSide,
+	surplusSide balanceRequirement) btcutil.Amount {
+
+	// Get the maximum shift allowed on our surplus side that is possible
+	// before it dips beneath its minimum.
+	available := surplusSide.currentAmount - surplusSide.minimumAmount
+
+	// Get the midpoint between our two minimum amounts, expressed from
+	// the perspective of the side that we currently have a deficit of
+	// liquidity.
+	midpoint := (deficitSide.minimumAmount +
+		(capacity - surplusSide.minimumAmount)) / 2
+
+	// Calculate the amount of balance we need to shift our
+	// current deficit side to reach this midpoint.
+	midpointDelta := midpoint - deficitSide.currentAmount
+
+	// If the amount we have available is less than the midpoint between
+	// our two minimums, we do not recommend any action. This may happen
+	// if a large portion of our balance is in a pending state.
+	// TODO(carla): do a best-effort split attempt here?
+	if available < midpointDelta {
+		return 0
+	}
+
+	// Otherwise, we have sufficient surplus to shift our deficit side all
+	// the way to the midpoint. We return the amount we need to reach this
+	// midpoint so that we do not overshoot and produce a deficit in the
+	// opposite direction.
+	return midpointDelta
```

Comment:
It still seems that the code in this file can be simplified further. I'd expect something like:
```
midpoint = (min_balance + max_balance) / 2
if balance < min_balance { loop_in(midpoint-balance)}
if balance > max_balance {loop_out(balance - midpoint)}

Reply:
Simplified this a bit, but we still do need to do the available amount checks. It's not the case that we can always just swap `midpoint-balance` because if we have some pending htlcs, this may tip us below our outgoing threshold anyway.